### PR TITLE
Development PR

### PR DESCRIPTION
**Pull request by mvasishta-ds**
This PR allows for review of communications middleware development.

Class diagram:
Blank diagram - UML Class

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+set(CMAKE_CXX_STANDARD 17)
+
+find_package(Protobuf REQUIRED)
+
+add_subdirectory(lib_comms)
+add_subdirectory(client)
+add_subdirectory(server)
+add_subdirectory(test)
+
+set(CMAKE_DISABLE_IN_SOURCE_BUILD ON)
```
Comment (Janet-DeepSight) [inline CMakeLists.txt:13]: FMI: What is an in source build?

**Comment by Janet-DeepSight** [comment]
Hi Manoj, can you add a bit more comments about what things are.  Maybe in the header files.  This is what I'm piecing together:
- everything in the zmp folder (ZmqPublisher, ZmqPubSubBackend, etc) are all wrappers around zmq - correct?
- There are some abstract classes to help create an interface to the backend implementation (PublisherImpl, PubSubBackend, SubscriberImpl).  Is this correct?
- client and server folder are both apps for testing.  Maybe should be moved to the test folder?  Adding comments about what they are would be helpful.

We are following the [Google C++ Style Guide](https://google.github.io/styleguide/cppguide.html) so may need to update variable naming convention when you merge into main:
- Filenames should be all lowercase and can include underscores (_) or dashes (-).
- Type names start with a capital letter and have a capital letter for each new word, with no underscores: MyExcitingClass, MyExcitingEnum.
- The names of variables (including function parameters) and data members are all lowercase, with underscores between words. Data members of classes (but not structs) additionally have trailing underscores. For instance: a_local_variable, a_struct_data_member, a_class_data_member_.
- ETC

I also created a [Clang Format] (https://github.com/deepsightinc/DSTMain/blob/Develop/.clang-format) that I think I am using correctly to be applied on save....

**Review by Janet-DeepSight — COMMENTED** [review]
Did an initial look through.  I'll try to add more later tonight and tomorrow.  Thank you!

File: client/client.cpp [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+    if(!ip_option->is_set() || !port_option->is_set() ) {
+        std::cout << "Error: Missing required arguments" << std::endl;
+        std::exit(0);
+    }
+
+    std::cout << "starting client with server IP address: " << ip_option->value() << " port: " << port_option->value() << std::endl;
+
+    Comms middleware;
+    auto subscriber = middleware.CreateSubscriber<Topic::BasicTopic>(ip_option->value(), port_option->value());
```
[the remark this answers is not in the document]
Comment (mvasishta-ds) [reply client/client.cpp:27]: That is correct, there is a 1-to-1 mapping from subscriber to topic. This was done mainly for implementation simplicity and robustness.

File: lib_comms/Topics.h
[the remark this answers is not in the document]
Comment (mvasishta-ds) [reply lib_comms/Topics.h:1]: I was intending that separate libraries would contain their own Topic definitions that work with other types of Protobuf payloads - the topic provided here is just a basic one. I can add more documentation here to clarify that.

File: CMakeLists.txt (hunk 1 above)
Comment (mvasishta-ds) [reply CMakeLists.txt:13]: in-source builds are where the binaries of the build outputs are located in the same directory as the source files. I find that its generally cleaner to have the build artifacts in a separate location (as is already done in DSTMain)

**Comment by mvasishta-ds** [comment]
Hi Janet,

Addressing each of your points:
* Everything in the zmq folder is indeed wrapper around zmq (specifically [this](https://github.com/zeromq/cppzmq) C++ API)
* The PublisherImpl and SubscriberImpl classes are indeed meant to be abstractions to support different underlying pub/sub technology. PubSubBackend is used to create PublisherImpl and SubscriberImpl and store shared data that is relevant. (For zmq, a shared threading context. For fakes, a mapping between topic and subscribers.)
* I imagined that during integration the client and server programs would be moved to a separate top-level "utilities" folder rather than being grouped together with the unit tests - thoughts?
* I will clang-format the files during integration
* I will add comments to the interface classes to better describe their responsibilities/functionalities

File: CMakeLists.txt (hunk 1 above)
Comment (Janet-DeepSight) [reply CMakeLists.txt:13]: I totally agree!  Didn't know there was a way to enforce it.  Nice.

**Comment by Janet-DeepSight** [comment]
[sc-7753]

**Review by Janet-DeepSight — APPROVED** [review]

**Pull request merged**